### DISCUSSION TITLE
If I use Atom on a mounted component, how can I avoid multiple reads when dependencies are updated?

### DISCUSSION THREAD

**Discussion by hi-ogasawara** [pmndrs/jotai#2782] [Q&A] [answered]
Hi
When using Atom on a mounted component
When you update a dependency, it will be read multiple times.
So I'm trying to avoid this by following these steps
Is there another way?
*I especially want to prevent multiple fetches.

step
 -  Set with atomWithDefault to cut dependencies
 -  Update dependencies within atomWithDefault
 -  RESET atomWithDefault to updated value

another idea(No Tried)
　Read dependent atoms and if the value has not changed, create an atom that returns early with the previously used value
　*It may be possible to be read multiple times but not be fetched multiple times

Tried and tested code
*Uncomment out console.log as needed to confirm

```typescript
import { atom , useAtom , useAtomValue , useSetAtom} from "jotai";
import { atomWithDefault , atomWithRefresh , RESET, useAtomCallback } from "jotai/utils";
import { useCallback } from "react";

const baseCount1 = atom(0);
const baseCount2 = atom(0);

const readCountAtom = atom((get) => {
  // console.log('readCountAtom:read');
  return get(baseCount1) + get(baseCount2);
})

const wrapReadCountDefault = atom((get) => {
  // console.log('wrapReadCountDefault:read');
  return get(readCountDefault);
});

const readCountDefault = atomWithDefault((get) => {
  console.log("readCountDefault:read");
  return get(baseCount1) + get(baseCount2);
});

const readCountRefresh = atomWithRefresh((get) => {
  // console.log("readCountRefresh:read")
  return get(baseCount1) + get(baseCount2);
});

function MountReadAtom() {
  const [countRefresh, setRefresh] = useAtom(readCountRefresh);
  const [countDefault, setReset] = useAtom(readCountDefault);
  const wrapCountDefault = useAtomValue(wrapReadCountDefault);
  const countAtom = useAtomValue(readCountAtom);

  const setCount1 = useSetAtom(baseCount1);
  const setCount2 = useSetAtom(baseCount2);

  // Read multiple times
  const handleRefresh = useCallback(() => {
    setCount1((prev) => prev + 1);
    setCount2((prev) => prev + 1);
    setRefresh();
  }, [setCount1, setCount2, setRefresh]);

  // Read once
  const handleDefault = useCallback(() => {
    setReset(0);
    setCount1((prev) => prev + 1);
    setCount2((prev) => prev + 1);
    setReset(RESET);
  }, [setCount1, setCount2, setReset]);

  // Read multiple times
  const handleCountAtom = useCallback(() => {
    setCount1((prev) => prev + 1);
    setCount2((prev) => prev + 1);
  },[setCount1, setCount2])

  // Read once
  const handleDefaultAtomCallBack = useAtomCallback(
    useCallback(() => {
      setReset(0);
      setCount1((prev) => prev + 1);
      setCount2((prev) => prev + 1);
      setReset(RESET);
    }, [setCount1, setCount2, setReset])
  );

  return (
    <>
      <button onClick={handleRefresh}>Refresh: {countRefresh}</button>
      <button onClick={handleDefault}>Default: {countDefault}</button>
      <button onClick={handleCountAtom}>commonAtom: {countAtom}</button>
      <button onClick={handleDefaultAtomCallBack}>
        handleDefaultAtomCallBack
      </button>
      <label>WrapDefault: {wrapCountDefault}</label>
    </>
  );
}

export default MountReadAtom;
```

**Comment by dai-shi** [comment]
I'm not sure if I follow 100%, but can you define a write-only atom to prevent it?

**Reply by hi-ogasawara** [reply]
@dai-shi 
Thank you for answering!

If I use get with a write-only atom in useEffect (side effects) or event handlers, it should work since dependencies won’t be tracked. However, for the following reasons, I have been using a read-only atom, so if there are any alternative methods.
I have written a [new comment](https://github.com/pmndrs/jotai/discussions/2782#discussioncomment-11069468) about the current situation and how to use it, and I would be very grateful if you could answer it for me.

Note: Regarding AtomCallback (write-only atom), I noticed some behavior that didn’t seem intended, so I plan to submit a separate [bug report](https://github.com/pmndrs/jotai/discussions/2789).

Reasons:
 - I used a read-only atom to retrieve data during rendering.
   I wanted to avoid implementing loading display logic myself, which would be necessary without setting it to a Promise type.

 - Sometimes, I want dependencies to be tracked (this is more important) .
   I want it to fetch and retrieve new data when dependencies change.  If I were to use a write-only atom, I would need to re-fetch in
   every event handler where dependencies might change, which I’d like to avoid. For example, when fetching page data, 
   I’d need to add re-fetch logic to every event handler where dependencies might change.

**Comment by hi-ogasawara** [comment]
@dai-shi 

There are several read-only atoms that communicate with the backend. They use multiple common atoms (dependencies) for fetching data. For example:

```typescript
export const fetch1 = atom(async (get) => {
  const atomData1 = await get(atom1);
  const atomData2 = await get(atom2);
  return await fetchData1(atomData1, atomData2);
});

export const fetch2 = atom(async (get) => {
  const atomData1 = await get(atom1);
  const atomData2 = await get(atom2);
  return await fetchData2(atomData1, atomData2);
});
```

It’s very convenient that updating `atom1` and `atom2` will cause `fetch1` and `fetch2` to be re-evaluated by Jotai, and I’d like to take advantage of this mechanism.

When `atom1` and `atom2` are updated, multiple fetches occur. I want to limit this to a single fetch. However, is it a specification that multiple reads occur when updating several dependent atoms?

If there are cases where an atom is read multiple times and others where it’s not, please let me know.🙇‍♂️

**Reply by dai-shi** [reply]
Hi, sorry for the delay.

copy-pasting your code into StackBlitz: https://stackblitz.com/edit/vitejs-vite-nalfnw?file=src%2FApp.tsx&terminal=dev

Is it too naive? It doesn't seem multiple fetches happen in this case. Can you modify it to show your problem? I wonder if Promise.all can be a solution.

**Reply by hi-ogasawara** [reply]
@dai-shi 
Thank you for trying this on StackBlitz!
I forked and added some patterns.
https://stackblitz.com/edit/vitejs-vite-si3fhc?file=src%2Fmain.tsx

Pattran1
If the order of atoms is changed,  it will be read multiple times.
The same is true if you change the order of updates in the event handler.
```typescript
const fetch1 = atom(async (get) => {
  console.log("fetch1:read")
  const atomData2 = await get(atom2);
  const atomData1 = await get(atom1);
  return await fetchData1(atomData1, atomData2);
});

const fetch2 = atom(async (get) => {
  console.log("fetch2:read")
  const atomData1 = await get(atom1);
  const atomData2 = await get(atom2);
  return await fetchData2(atomData1, atomData2);
});
```

Pattran2
If updated synchronously, it will be read multiple times.
```typescript
const fetch1 = atom(async (get) => {
  console.log("fetch1:read")
  const atomData2 = get(atom2);
  const atomData1 = get(atom1);
  return await fetchData1(atomData1, atomData2);
});

const fetch2 = atom(async (get) => {
  console.log("fetch2:read")
  const atomData1 = get(atom1);
  const atomData2 = get(atom2);
  return await fetchData2(atomData1, atomData2);
});
```

Pattran3
If initially synchronous and later asynchronous, it is read multiple times.
```typescript
const fetch1 = atom(async (get) => {
  console.log('fetch1:read');
  const atomData1 = get(atom1);
  const atomData2 = await get(atom2);
  return await fetchData1(atomData1, atomData2);
});

const fetch2 = atom(async (get) => {
  console.log('fetch2:read');
  const atomData1 = get(atom1);
  const atomData2 = await get(atom2);
  return await fetchData2(atomData1, atomData2);
});
```

Pattern4
If asynchronous from the beginning, it is read only once, but if the order of atoms is changed, it is read multiple times.
```typescript
const fetch1 = atom(async (get) => {
  console.log("fetch1:read")
  const atomData1 = await get(atom1);
  const atomData2 = get(atom2);
  return await fetchData1(atomData1, atomData2);
});

const fetch2 = atom(async (get) => {
  console.log("fetch2:read")
  const atomData3 = await get(atom3);
  const atomData2 = get(atom2);
  return await fetchData2(atomData3, atomData2);
});
```

Pattern5
Even if Promise.all is used, it is read multiple times.
```typescript
const fetch1 = atom(async (get) => {
  console.log("fetch1:read")
  const [ atomData1 , atomData2] = await Promise.all([get(atom1) , get(atom2)]);
  return await fetchData1(atomData1, atomData2);
});

const fetch2 = atom(async (get) => {
  console.log("fetch2:read")
  const [ atomData2 , atomData1] = await Promise.all([get(atom2) , get(atom1)]);
  return await fetchData2(atomData1, atomData2);
});
```

Summary.
If the first atom is of Promise type (await get(atom)) and the order of updating dependencies is the same, it seems possible to handle the dependency changes so that they are not read multiple times.

**Reply by dai-shi** [reply]
Thanks for the examples.
Yeah, this is unavoidable. Two updates might happen with some delays.
One solution is to use `signal`, which works in this case.
https://stackblitz.com/edit/vitejs-vite-bj5yl3?file=src%2FPattern1.tsx

Another solution would be to use "debounce".

**Reply by hi-ogasawara** [reply]
@dai-shi 
Thank you! I now have a better understanding.
Why does signal not read multiple times?
Are you waiting for an update on the dependent atom to abort?

**Reply by dai-shi** [reply]
We abort old execution when dependencies change.

**Reply by hi-ogasawara** [reply]
@dai-shi 
I see.
Is it OK to use signal.aborted just to solve the problem of multiple reads?

**Reply by dai-shi** [reply]
`signal` is designed for such cases, so using it is fine.

I just converted the example to avoid "async" completely, 
https://stackblitz.com/edit/vitejs-vite-wjefeq?file=src%2FApp.tsx
but the problem still persists. I thought this case was already supported.
cc @dmaskasky

**Reply by hi-ogasawara** [reply]
@dai-shi 
I misunderstood that you were talking about `signal`
Sorry for the late reply.

The above link is the pattern with async removed from fetch1 and fetch2, correct?
That was what I wanted to try with [Pattran2](https://github.com/pmndrs/jotai/discussions/2782#discussioncomment-11183746). My mistake.

```typescript
const fetch1 = atom((get) => {
  console.log("fetch1:read")
  const atomData2 = get(atom2);
  const atomData1 = get(atom1);
  return await fetchData1(atomData1, atomData2);
});

const fetch2 = atom((get) => {
  console.log("fetch2:read")
  const atomData1 = get(atom1);
  const atomData2 = get(atom2);
  return await fetchData2(atomData1, atomData2);
});
```

Does the above synchronous update of multiple dependencies mean that it should be read only once, but is being read multiple times?

**Reply by dai-shi** [reply]
I thought we made it so that it would call just once. But, I might be missing something. So, it feels like a bug in the sync case.

**Reply by dmaskasky** [reply, marked as answer]
In Jotai, `set` recomputes the dependents of an atom in place. This ensures that if those values are accessed in the very next line using `get`, they reflect the updated state. Consider the following example:

```ts
const a = atom(0);
const b = atom((get) => get(a) + 1);
const w = atom(null, (get, set) => {
  const b1 = get(b); // 1
  set(a, (v) => v + 1); // Recompute dependents of `a` here
  const b2 = get(b); // 2
});
```

This behavior means that every time an atom's value is updated with `set`, all its dependents are recalculated, even if they aren't accessed in the write function.

_Edit:_ I am exploring changing this behavior so that only the dependents accessed with `get` will be recomputed in place, all others are deferred until the end. https://github.com/pmndrs/jotai/pull/2827

**Reply by dai-shi** [reply]
To recap, 

> I thought we made it so that it would call just once.

It was my misunderstanding.
